### PR TITLE
Topic recognition

### PR DESCRIPTION
**Pull request by peterdbeardsley**
Hi Shakes,

My pull request provides a solution to task number 9 - a generative model of the OASIS brain dataset using StyleGAN. This code will synthesize new OASIS brain scans for brain images as well as segmentations at a resolution of 256x256 and 512x512.

In my solution, I followed the StyleGAN architecture as best as I could. The architecture comprises of a mapping network of fully connected layers that learns a manifold, w, for the latent z. The w weights are used as a scale and bias to style each progressive GAN synthesis layer via an Adaptive Instance Normalisation class (AdaIN). I also included random noise inputs prior to each AdaIN to facilitate stocastic behaviour that facilitates variation in details of the synthesized images. The first synthesis layer begins with a 4x4 constant image. Each subsequent layer doubles the image size until the original 256x256 image size is obtained. I also experimented with supersized images of 512x512. I had to experiment a lot with the loss function. After many failures, I settled on an already established approach that uses gradient penalty, Wasserstein distance, and gradient clipping.

The StyleGAN results look very promising, and I believe they are a "reasonably clear image". For me it is difficult to distinguish the difference between a fake and real image about half the time. In the other half, there is minor blurriness and lack of crisp detail that gives away the fakes.

I developed, tested, and trained the model using PyTorch and CUDA for version 2.0.1, and Torchvision version 0.15.2.

Regards,

Peter Beardsley (s45095904)

### CODE REVIEW

**Comment by WeiDaiwow** [comment]
# Document

- Title: 
	Good
- Address probelm:
	Good
- Working Principle:
	Good
- Figure visualization
	Good
- dependency
	Good
- inputs and outputs description:
	Good
- training plot:
	Good
- readme using markdown:
	Good
- preprocessing description:
	Good
- dataset split description
	Missing
- File description:
	Good

Problem:
- The training between D and G does not seem to be balanced. The D loss became really large at end of the training. It would be good to evaluate such problem.  EG. what happened to the generated images after 80000 iterations.

# Check the files:
- modules.py
	Good
- dataset.py
	Good
- train.py
	Good
- predict.py
	Good
Minor tips (nothing to do with marking)
1. During model training, it is better to zero_grad() the optimizer instead of the model itself.


# Overall:

- solve the problem:
	Yes
- TF/torch implementation:
	Yes
- Design quality:
	Good
- code comments:
	Good
- Difficulty:
	Hard -SytleGANv1

Well Done

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe good, no usage -1
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good

**Comment by shakes76** [comment]
Please remove gitignore file to allow merge. Does not affect grade, just for merging.

**Comment by peterdbeardsley** [comment]
Physically deleted .gitignore

**Comment by wangzhaomxy** [comment]
Feedback attempt and no feedback marks lost.
